frontend/catalyst/compiler.py

```diff
@@ -264,233 +383,111 @@ def link(infile, outfile, flags=None, fallback_compilers=None, compile_options=N
-        Optional compile_options (CompileOptions): generic compilation options.
-    Returns:
-        a path to the output file
-    """
-    if filename[-10:] != ".llvm.mlir":
-        raise ValueError(
-            f"Input file ({filename}) for LLVMIR conversion is not an LLVM dialect MLIR file"
-        )
-
-    new_fname = filename.replace(".llvm.mlir", ".ll")
-
-    command = [translate_tool]
-    command += [filename]
-    command += ["--mlir-to-llvmir"]
-    command += ["-o", new_fname]
-
-    run_writing_command(command, compile_options)
-
-    return new_fname
-
-
-def compile_llvmir(filename: str, compile_options: Optional[CompileOptions] = None) -> str:
-    """Translate LLVM IR to an object file.
-
-    Args:
-        filename (str): the path to a file were the program is stored.
-        Optional compile_options (CompileOptions): generic compilation options.
-    Returns:
-        a path to the output file
-    """
-    if filename[-3:] != ".ll":
-        raise ValueError(f"Input file ({filename}) for compilation is not an LLVMIR file")
-
-    new_fname = filename.replace(".ll", ".o")
-
-    command = [compiler]
-    command += compiler_flags
-    command += [filename]
-    command += ["-o", new_fname]
-
-    run_writing_command(command, compile_options)
-
-    return new_fname
-
-
-def link_lightning_runtime(filename: str, compile_options: Optional[CompileOptions] = None) -> str:
-    """Link the object file as a shared object.
-
-    Args:
-        filename (str): the path to a file were the object file is stored.
-        Optional compile_options (CompileOptions): generic compilation options.
-    Returns:
-        a path to the output file
-    """
-    if filename[-2:] != ".o":
-        raise ValueError(f"Input file ({filename}) for linking is not an object file")
+        Args:
+            mlir_module (Module): the MLIR module
+            pipelines (List[Any]): the list of compilation pipelines
 
-    new_fname = filename.replace(".o", ".so")
+        Returns:
+            Shared object
+        """
+        module_name = mlir_module.operation.attributes["sym_name"]
+        # Convert MLIR string to Python string
+        module_name = str(module_name)
+        # Remove quotations
+        module_name = module_name.replace('"', "")
+
+        if keep_intermediate:
+            parent_dir = os.getcwd()
+            path = os.path.join(parent_dir, module_name)
+            os.makedirs(path, exist_ok=True)
+            workspace_name = os.path.abspath(path)
+        else:
+            workspace_name = self.workspace.name
 
-    CompilerDriver.link(filename, new_fname, compile_options=compile_options)
+        if pipelines is None:
+            pipelines = [
+                MHLOPass,
+                QuantumCompilationPass,
+                BufferizationPass,
+                MLIRToLLVMDialect,
+                LLVMDialectToLLVMIR,
+                LLVMIRToObjectFile,
+                CompilerDriver,
+            ]
```

Comment:
It would be nice to update the compiler architectural diagram based on this refactoring. This can be simply done by updating `compiler.puml`. Let me know if you need a help :100:

Reply:
Awesome! :+1: